### PR TITLE
Refactor movement

### PR DESCRIPTION
**Pull request by BlairArchibald**
This PR refactors movement based on some of the changes Peter showed in https://github.com/ScottishCovidResponse/Covid_Simulation_Model/pull/86

Importantly it makes it harder to directly access people, and moves movement functionality into Person where it can be extended to support public transport etc in future.

I've also added a test to check if we accidentally allow duplicate people anywhere.

### CODE REVIEW

File: src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java [hunk 1]
```diff
@@ -65,17 +66,16 @@ public void setSize(Size s) {
...
             }
 
             if (!p.worksNextHour(this, t, lockdown)) {
-                p.returnHome();
                 left.add(p);
             }
         }
-        people.removeAll(left);
+        left.forEach(p -> p.returnHome(this));
     }
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java:79]: Would this (simpler) approach work instead?
```java
    public void moveShifts(Time t, boolean lockdown, Function<Person, Boolean> filter) {
        List<Person> peopleWhoMightGoHome = new ArrayList<>(people);
        for (Person p : peopleWhoMightGoHome) {
            if (filter.apply(p)) {
                continue;
            }

            if (!p.worksNextHour(this, t, lockdown)) {
                p.returnHome();
            }
        }
    }
```

**Comment by BlairArchibald** [comment]
I've added a first pass at a slightly different movement system where:

- We control access to people by forcing getPeople to only allow iterating over it (and a couple of non-modifiable helpers). There's still add/remove person but these should only be used at the start to allocate households etc.
- **Everyone** must move each timestep, even if this move is to stay put. Only nextPeople is modified and each person tracks the movement (we still have a *from* that's not used yet to handle public transport)

This reduces allocations, avoids needing copies of people, allows people to be GC'd in 1 step (if required; currently we just clear it and there's not much different performance wise).

There's probably some work to do to streamline some of the filtering etc. I'm still not 100% convinced by some of the functions that need to check if they should be at work again (when we've already checked this). But the basic outline is here.

Profiling shows this to be very slightly faster than before (about 4s saving on 60 days with 100k people).

Note: We might move back to people as a directly accessible list at some point since it seems for (Person p : getPeople()) can be slower than for (int ...) [extra bounds checks or something]

**Comment by johnnonweiler** [comment]
Thank you.  This basic outline is looking a lot better.
(I'm surprised you see a noticable performance hit for iterating through a list.)

**Comment by BlairArchibald** [comment]
There's probably a bit more work we can do on the movement system (possibly to reduce the number of times we need to iterate through people), but I think it's in a decent place just now so I'm tempted to merge this now to allow easier implementation of public transport/hospital visists/lockdowns etc, and continue refactoring as we go along.
